### PR TITLE
Enhancement #146 add a thisID and thisButton

### PR DESCRIPTION
**Pull request by shaunhillier**
Added a thisID variable to default Variable object.
Added thisButton to default collections object.

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3175/pr-test (or any other room on that server)

**Comment by shaunhillier** [comment]
I'm quite new to GITHub, so probably didn't need 11 commits. If there is a better way to do the pull request let me know.

File: client/js/widgets/button.js
Comment (ArnoldSmith86) [inline client/js/widgets/button.js:56]: Any reason for removing the empty line?

File: client/js/widgets/button.js
Comment (ArnoldSmith86) [inline client/js/widgets/button.js:56]: The indentation is missing a space.

This can also be `const` because it will never change.

File: client/js/widgets/button.js
Comment (ArnoldSmith86) [inline client/js/widgets/button.js:60]: This is actually the same as writing `thisID: thisID` which in turn can be shortened to just `thisID` (like the others).

Since you don't need `thisID` anywhere else (see comment below), I would have probably written `thisID: this.p('id')` but that's just me.

File: client/js/widgets/button.js
Comment (ArnoldSmith86) [inline client/js/widgets/button.js:65]: You obviously copied this from the other code that creates a collection. Which is a reasonable thing to do.

The parenthesis around `Array.from` are not necessary.

`if(w.p("id")==thisID) return true;` could just be `return w.p("id")==thisID;`. The whole `function(w) { return w.p("id")==thisID; }` can also be written as `w=>w.p("id")==thisID`. You don't even need to look at the ID though. A `w==this` would be fine.

In the end you're doing all of that to go through all the `widgets` and only return the one you're currently handling. So in the end you could just write `const thisButton = [ this ];` because you already know exactly what should be in the array.

I'm guessing you weren't sure if the `widgets` array contains the same stuff as the `this`. It does.

File: client/js/widgets/button.js
Comment (ArnoldSmith86) [inline client/js/widgets/button.js:67]: Same as above. Could just be `thisButton: [ this ]`.

**Review by ArnoldSmith86 — CHANGES_REQUESTED** [review]
I did not test your code but I'm pretty sure it works.

You already said in Discord that you realized too late that you can test without committing so that's why there are 11 commits here. No problem.

Please do write proper commit messages though. In the end this should have been one commit with a message similar to "added thisID and thisButton as presets to button variables".


In general, thank you very much for getting into the code! Consider most of this feedback only as hints because the code is (probably) working fine. I feel like for Javascript it's especially difficult to find proper examples online because it evolved so much over the decades.

File: client/js/widgets/button.js
Comment (shaunhillier) [reply client/js/widgets/button.js:65]: You've hit the nail on the head, I didn't realise they were exactly the same

File: client/js/widgets/button.js
Comment (shaunhillier) [reply client/js/widgets/button.js:56]: No reason, I originally had a lot of logs to see what different variables held, and when I removed them I must have removed the extra line

**Review by RaphaelAlvez — APPROVED** [review]
Tested and seems to be working

**Review by 96LawDawg — APPROVED** [review]
Concur.  Test works.  I know my second review no longer means approval, but at least Arnold as codeowner knows there was another look.

File: client/js/widgets/button.js
Comment (ArnoldSmith86) [inline client/js/widgets/button.js:64]: Please keep indentation consistent. :)

**Review by ArnoldSmith86 — CHANGES_REQUESTED** [review]

File: client/js/widgets/button.js
Comment (shaunhillier) [reply client/js/widgets/button.js]: I couldn't even see the issue at first :) but resolved now.

**Review by ArnoldSmith86 — APPROVED** [review]
Code looks good. Didn't test it though.

**Pull request merged**
